lib/modules.nix

```diff
@@ -869,9 +869,17 @@ rec {
   # option declarations in submodules have accurate position information.
   # TODO: Merge this into mergeOptionDecls
   fixupOptionType = loc: opt:
-    if opt.type.getSubModules or null == null
+    builtins.addErrorContext ''
+      while evaluating the type of option `${showOption loc}`
+
+      ====================================
+      A possible cause for this error is that `${showOption loc}` has a recursively
```

Comment:
It definitely gets detected, I used this error to find what I needed to fix. I think it's because `getSubModules` is not a function? See the full trace:

<details>
<summary>Details</summary>

```
error: infinite recursion encountered

       at /home/n/git/nixpkgs/lib/types.nix:179:14:

          178|     { _type = "option-type";
          179|       inherit
             |              ^
          180|         name check merge emptyValue getSubOptions getSubModules substSubModules

       … while evaluating the attribute 'getSubModules'

       at /home/n/git/nixpkgs/lib/types.nix:179:14:

          178|     { _type = "option-type";
          179|       inherit
             |              ^
          180|         name check merge emptyValue getSubOptions getSubModules substSubModules

       … while evaluating the attribute 'getSubModules'

       at /home/n/git/nixpkgs/lib/types.nix:179:14:

          178|     { _type = "option-type";
          179|       inherit
             |              ^
          180|         name check merge emptyValue getSubOptions getSubModules substSubModules

       … while evaluating the attribute 'getSubModules'

       at /home/n/git/nixpkgs/lib/types.nix:179:14:

          178|     { _type = "option-type";
          179|       inherit
             |              ^
          180|         name check merge emptyValue getSubOptions getSubModules substSubModules

       … while evaluating the attribute 'getSubModules'

       at /home/n/git/nixpkgs/lib/types.nix:179:14:

          178|     { _type = "option-type";
          179|       inherit
             |              ^
          180|         name check merge emptyValue getSubOptions getSubModules substSubModules

       … while evaluating the attribute 'type.getSubModules'

       at /home/n/git/nixpkgs/lib/types.nix:179:14:

          178|     { _type = "option-type";
          179|       inherit
             |              ^
          180|         name check merge emptyValue getSubOptions getSubModules substSubModules

       … while evaluating the type of option `services.yggdrasil.settings`

       ====================================
       A possible cause for this error is that `services.yggdrasil.settings` has a recursively
       defined type that does not use `types.recursive`.
       ====================================


       … while calling 'fixupOptionType'

       at /home/n/git/nixpkgs/lib/modules.nix:871:26:

          870|   # TODO: Merge this into mergeOptionDecls
          871|   fixupOptionType = loc: opt:
             |                          ^
          872|     builtins.addErrorContext ''

       … from call site

       at /home/n/git/nixpkgs/lib/modules.nix:591:23:

          590|           if length optionDecls == length decls then
          591|             let opt = fixupOptionType loc (mergeOptionDecls loc decls);
             |                       ^
          592|             in {

       … while calling 'warnIf'

       at /home/n/git/nixpkgs/lib/trivial.nix:357:18:

          356|   */
          357|   warnIf = cond: msg: if cond then warn msg else x: x;
             |                  ^
          358|

       … from call site

       at /home/n/git/nixpkgs/lib/modules.nix:719:9:

          718|       warnDeprecation =
          719|         warnIf (opt.type.deprecationMessage != null)
             |         ^
          720|           "The type `types.${opt.type.name}' of option `${showOption loc}' defined in ${showFiles opt.declarations} is deprecated. ${opt.type.deprecationMessage}";

       … while calling 'evalOptionValue'

       at /home/n/git/nixpkgs/lib/modules.nix:694:31:

          693|      config value. */
          694|   evalOptionValue = loc: opt: defs:
             |                               ^
          695|     let

       … from call site

       at /home/n/git/nixpkgs/lib/modules.nix:593:32:

          592|             in {
          593|               matchedOptions = evalOptionValue loc opt defns';
             |                                ^
          594|               unmatchedDefns = [];

       … while evaluating the attribute 'matchedOptions'

       at /home/n/git/nixpkgs/lib/modules.nix:593:15:

          592|             in {
          593|               matchedOptions = evalOptionValue loc opt defns';
             |               ^
          594|               unmatchedDefns = [];

       … while calling anonymous lambda

       at /home/n/git/nixpkgs/lib/modules.nix:623:37:

          622|
          623|       matchedOptions = mapAttrs (n: v: v.matchedOptions) resultsByName;
             |                                     ^
          624|

       … from call site

       … while calling 'attrByPath'

       at /home/n/git/nixpkgs/lib/attrsets.nix:36:5:

           35|     # The nested attribute set to select values from
           36|     set:
             |     ^
           37|     let attr = head attrPath;

       … from call site

       at /home/n/git/nixpkgs/lib/attrsets.nix:41:12:

           40|       else if set ? ${attr}
           41|       then attrByPath (tail attrPath) default set.${attr}
             |            ^
           42|       else default;

       … while calling 'attrByPath'

       at /home/n/git/nixpkgs/lib/attrsets.nix:36:5:

           35|     # The nested attribute set to select values from
           36|     set:
             |     ^
           37|     let attr = head attrPath;

       … from call site

       at /home/n/git/nixpkgs/lib/attrsets.nix:41:12:

           40|       else if set ? ${attr}
           41|       then attrByPath (tail attrPath) default set.${attr}
             |            ^
           42|       else default;

       … while calling 'attrByPath'

       at /home/n/git/nixpkgs/lib/attrsets.nix:36:5:

           35|     # The nested attribute set to select values from
           36|     set:
             |     ^
           37|     let attr = head attrPath;

       … from call site

       at /home/n/git/nixpkgs/lib/attrsets.nix:41:12:

           40|       else if set ? ${attr}
           41|       then attrByPath (tail attrPath) default set.${attr}
             |            ^
           42|       else default;

       … while calling 'attrByPath'

       at /home/n/git/nixpkgs/lib/attrsets.nix:36:5:

           35|     # The nested attribute set to select values from
           36|     set:
             |     ^
           37|     let attr = head attrPath;

       … from call site

       at /home/n/git/nixpkgs/lib/modules.nix:1147:26:

         1146|         (abort "Renaming error: option `${showOption to}' does not exist.");
         1147|       toType = let opt = attrByPath to {} options; in opt.type or (types.submodule {});
             |                          ^
         1148|     in

       … while calling 'optionalAttrs'

       at /home/n/git/nixpkgs/lib/attrsets.nix:610:5:

          609|     # The attribute set to return if `cond` is `true`.
          610|     as:
             |     ^
          611|     if cond then as else {};

       … from call site

       at /home/n/git/nixpkgs/lib/modules.nix:1154:12:

         1153|         apply = x: use (toOf config);
         1154|       } // optionalAttrs (toType != null) {
             |            ^
         1155|         type = toType;

       … while calling 'atDepth'

       at /home/n/git/nixpkgs/lib/attrsets.nix:85:17:

           84|       len = length attrPath;
           85|       atDepth = n:
             |                 ^
           86|         if n == len

       … from call site

       at /home/n/git/nixpkgs/lib/attrsets.nix:88:39:

           87|         then value
           88|         else { ${elemAt attrPath n} = atDepth (n + 1); };
             |                                       ^
           89|     in atDepth 0;

       … while evaluating the attribute 'options'

       at /home/n/git/nixpkgs/lib/modules.nix:555:38:

          554|       declsByName = byName "options" (module: option:
          555|           [{ inherit (module) _file; options = option; }]
             |                                      ^
          556|         ) options;

       … while calling 'isType'

       at /home/n/git/nixpkgs/lib/types.nix:70:18:

           69| rec {
           70|   isType = type: x: (x._type or "") == type;
             |                  ^
           71|

       … from call site

       at /home/n/git/nixpkgs/lib/modules.nix:588:36:

          587|           defns' = defnsByName'.${name} or [];
          588|           optionDecls = filter (m: isOption m.options) decls;
             |                                    ^
          589|         in

       … while calling anonymous lambda

       at /home/n/git/nixpkgs/lib/modules.nix:588:33:

          587|           defns' = defnsByName'.${name} or [];
          588|           optionDecls = filter (m: isOption m.options) decls;
             |                                 ^
          589|         in

       … from call site

       … while calling anonymous lambda

       at /home/n/git/nixpkgs/lib/modules.nix:582:39:

          581|
          582|       resultsByName = mapAttrs (name: decls:
             |                                       ^
          583|         # We're descending into attribute ‘name’.

       … from call site

       … while calling anonymous lambda

       at /home/n/git/nixpkgs/lib/modules.nix:628:22:

          627|         # Propagate all unmatched definitions from nested option sets
          628|         mapAttrs (n: v: v.unmatchedDefns) resultsByName
             |                      ^
          629|         # Plus the definitions for the current prefix that don't have a matching option

       … from call site

       … while calling anonymous lambda

       at /home/n/git/nixpkgs/lib/modules.nix:642:46:

          641|         else
          642|           concatLists (mapAttrsToList (name: defs:
             |                                              ^
          643|             map (def: def // {

       … from call site

       at /home/n/git/nixpkgs/lib/attrsets.nix:479:16:

          478|     attrs:
          479|     map (name: f name attrs.${name}) (attrNames attrs);
             |                ^
          480|

       … while calling anonymous lambda

       at /home/n/git/nixpkgs/lib/attrsets.nix:479:10:

          478|     attrs:
          479|     map (name: f name attrs.${name}) (attrNames attrs);
             |          ^
          480|

       … from call site

       … while evaluating the attribute 'unmatchedDefns'

       at /home/n/git/nixpkgs/lib/modules.nix:635:7:

          634|       # Transforms unmatchedDefnsByName into a list of definitions
          635|       unmatchedDefns =
             |       ^
          636|         if configs == []

       … while calling anonymous lambda

       at /home/n/git/nixpkgs/lib/modules.nix:628:22:

          627|         # Propagate all unmatched definitions from nested option sets
          628|         mapAttrs (n: v: v.unmatchedDefns) resultsByName
             |                      ^
          629|         # Plus the definitions for the current prefix that don't have a matching option

       … from call site

       … while calling anonymous lambda

       at /home/n/git/nixpkgs/lib/modules.nix:642:46:

          641|         else
          642|           concatLists (mapAttrsToList (name: defs:
             |                                              ^
          643|             map (def: def // {

       … from call site

       at /home/n/git/nixpkgs/lib/attrsets.nix:479:16:

          478|     attrs:
          479|     map (name: f name attrs.${name}) (attrNames attrs);
             |                ^
          480|

       … while calling anonymous lambda

       at /home/n/git/nixpkgs/lib/attrsets.nix:479:10:

          478|     attrs:
          479|     map (name: f name attrs.${name}) (attrNames attrs);
             |          ^
          480|

       … from call site

       … while evaluating the attribute 'unmatchedDefns'

       at /home/n/git/nixpkgs/lib/modules.nix:635:7:

          634|       # Transforms unmatchedDefnsByName into a list of definitions
          635|       unmatchedDefns =
             |       ^
          636|         if configs == []

       … while calling anonymous lambda

       at /home/n/git/nixpkgs/lib/modules.nix:628:22:

          627|         # Propagate all unmatched definitions from nested option sets
          628|         mapAttrs (n: v: v.unmatchedDefns) resultsByName
             |                      ^
          629|         # Plus the definitions for the current prefix that don't have a matching option

       … from call site

       … while calling anonymous lambda

       at /home/n/git/nixpkgs/lib/modules.nix:642:46:

          641|         else
          642|           concatLists (mapAttrsToList (name: defs:
             |                                              ^
          643|             map (def: def // {

       … from call site

       at /home/n/git/nixpkgs/lib/attrsets.nix:479:16:

          478|     attrs:
          479|     map (name: f name attrs.${name}) (attrNames attrs);
             |                ^
          480|

       … while calling anonymous lambda

       at /home/n/git/nixpkgs/lib/attrsets.nix:479:10:

          478|     attrs:
          479|     map (name: f name attrs.${name}) (attrNames attrs);
             |          ^
          480|

       … from call site

       … while evaluating the attribute 'unmatchedDefns'

       at /home/n/git/nixpkgs/lib/modules.nix:635:7:

          634|       # Transforms unmatchedDefnsByName into a list of definitions
          635|       unmatchedDefns =
             |       ^
          636|         if configs == []

       … while evaluating the attribute 'config'

       at /home/n/git/nixpkgs/lib/modules.nix:341:9:

          340|         options = checked options;
          341|         config = checked (removeAttrs config [ "_module" ]);
             |         ^
          342|         _module = checked (config._module);

       … while calling anonymous lambda

       at /home/n/git/nixpkgs/nixos/release.nix:143:48:

          142|
          143|   manualHTML = buildFromConfig ({ ... }: { }) (config: config.system.build.manual.manualHTML);
             |                                                ^
          144|   manual = manualHTML; # TODO(@oxij): remove eventually

       … from call site

       at /home/n/git/nixpkgs/nixos/release.nix:105:67:

          104|
          105|   buildFromConfig = module: sel: forAllSystems (system: hydraJob (sel (import ./lib/eval-config.nix {
             |                                                                   ^
          106|     inherit system;

       … while calling 'hydraJob'

       at /home/n/git/nixpkgs/lib/customisation.nix:221:14:

          220|      garbage collection. */
          221|   hydraJob = drv:
             |              ^
          222|     let

       … from call site

       at /home/n/git/nixpkgs/nixos/release.nix:105:57:

          104|
          105|   buildFromConfig = module: sel: forAllSystems (system: hydraJob (sel (import ./lib/eval-config.nix {
             |                                                         ^
          106|     inherit system;

       … while calling anonymous lambda

       at /home/n/git/nixpkgs/nixos/release.nix:105:49:

          104|
          105|   buildFromConfig = module: sel: forAllSystems (system: hydraJob (sel (import ./lib/eval-config.nix {
             |                                                 ^
          106|     inherit system;

       … from call site

       at /home/n/git/nixpkgs/lib/attrsets.nix:553:43:

          552|     f:
          553|     listToAttrs (map (n: nameValuePair n (f n)) names);
             |                                           ^
          554|
```

</details>